src/passes/Unsubtyping.cpp

```diff
@@ -0,0 +1,580 @@
+// because the value of type (ref $bot) no longer inhabits (ref $mid). To
+// prevent the cast's behavior from changing, we need to ensure that $bot <:
+// $mid.
+//
+// The set of subtyping requirements generated by a cast from $src to $dest is
+// that for every known remaining subtype $v of $src, if $v <: $dest in the
+// original module, then $v <: $dest in the optimized module. In other words,
+// for every type $v of values we know can flow into the cast, if the cast would
+// have succeeded for values of type $v before, then we know the cast must
+// continue to succeed for values of type $v. These requirements arising from
+// casts can also generate transitive requirements because we learn about new
+// types of values that can flow into casts as we learn about new subtypes of
+// cast sources.
+//
+// Starting with the initial subtype relation determined by walking the IR,
+// repeatedly search for new subtypings by analyzing type definitions and casts
+// in lock step until we reach a fixed point. This is the minimal subtype
+// relation that preserves module validity and behavior that can be found
+// without a more precise analysis of types that might flow into each cast.
+
+namespace wasm {
+
+namespace {
+
+struct Unsubtyping
+  : WalkerPass<ControlFlowWalker<Unsubtyping, OverriddenVisitor<Unsubtyping>>> {
+  // The new set of supertype relations.
+  std::unordered_map<HeapType, HeapType> supertypes;
+
+  // Map from cast source types to their destinations.
+  std::unordered_map<HeapType, std::unordered_set<HeapType>> castTypes;
+
+  // The set of subtypes that need to have their type definitions analyzed to
+  // transitively find other subtype relations they depend on. We add to it
+  // every time we find a new subtype relationship we need to keep.
+  std::unordered_set<HeapType> worklist;
+
+  void run(Module* wasm) override {
+    if (!wasm->features.hasGC()) {
+      return;
+    }
+    analyzePublicTypes(*wasm);
+    walkModule(wasm);
+    analyzeTransitiveDependencies();
+    optimizeTypes(*wasm);
+    // Cast types may be refinable if their source and target types are no
+    // longer related. TODO: Experiment with running this only after checking
+    // whether it is necessary.
+    ReFinalize().run(getPassRunner(), wasm);
+  }
+
+  void noteSubtype(HeapType sub, HeapType super) {
+    if (sub == super || sub.isBottom() || super.isBottom()) {
+      return;
+    }
+
+    auto [it, inserted] = supertypes.insert({sub, super});
+    if (inserted) {
+      worklist.insert(sub);
+      // TODO: Incrementally check all subtypes (inclusive) of sub against super
+      // and all its supertypes if we have already analyzed casts.
+      return;
+    }
+    // We already had a recorded supertype. The new supertype might be deeper,
+    // shallower, or identical to the old supertype.
+    auto oldSuper = it->second;
+    if (super == oldSuper) {
+      return;
+    }
+    if (HeapType::isSubType(super, oldSuper)) {
```

Comment:
Might be worth a comment here, I had to work it out in my head, that since each type has a single super, if we know a type is a subtype of two others then all three must form a chain and the two supers must be related as well.